e2e/daily-coding-challenge.spec.ts

```diff
@@ -0,0 +1,212 @@
+import {
+  getTodayUsCentral,
+  formatDate
+} from '../client/src/components/daily-coding-challenge/helpers';
+
+const dateRouteRe = /.*\/daily-coding-challenge\/date\/.*/;
+const allRouteRe = /.*\/daily-coding-challenge\/all/;
+
+const todayUsCentral = getTodayUsCentral();
+const [year, month, day] = todayUsCentral.split('-').map(Number);
+
+const todayMidnight = `${todayUsCentral}T00:00:00.000Z`;
+
+const mockApiChallenge = {
+  id: 'test-challenge-id',
+  challengeNumber: 1,
+  title: 'Test title',
+  date: todayMidnight,
+  description: 'Test description',
+  javascript: {
+    tests: [
+      {
+        text: 'Test text',
+        testString: 'assert.strictEqual(true, true);'
+      }
+    ],
+    challengeFiles: [
+      {
+        fileKey: 'scriptjs',
+        contents: '// JavaScript seed code'
+      }
+    ]
+  },
+  python: {
+    tests: [
+      {
+        text: 'Test text',
+        testString: '({test: () => { runPython(`assert True == True`)}})'
+      }
+    ],
+    challengeFiles: [
+      {
+        fileKey: 'mainpy',
+        contents: '# Python seed code'
+      }
+    ]
+  }
+};
+
+const mockApiAllChallenges = [
+  // today
+  {
+    // real ID from certified user so it shows completed in calendar
+    id: '6814d8e1516e86b171929de4',
+    date: todayMidnight
+  },
+  // yesterday, or tomorrow if today is the first
+  {
+    id: 'other-id',
+    date: `${formatDate({ year, month, day: day === 1 ? day + 1 : day - 1 })}T00:00:00.000Z`
+  }
+];
+
+const mockDaysInMonth = new Date(year, month, 0).getDate();
+
+test.describe('Daily Coding Challenges', () => {
+  test('should show not found page for invalid date', async ({ page }) => {
+    await page.goto('/learn/daily-coding-challenge?date=invalid-date');
+    await expect(
+      page.getByText(/daily coding challenge not found\./i)
+    ).toBeVisible();
+  });
+
+  test('should show not found page for date without challenge', async ({
+    page
+  }) => {
+    await page.route(dateRouteRe, async route => {
+      await route.fulfill({
+        status: 404,
+        headers: { 'Content-Type': 'application/json' },
+        json: { type: 'error', message: 'Challenge not found.' }
+      });
+    });
+
+    await page.goto('/learn/daily-coding-challenge?date=2025-01-01');
+    await expect(
+      page.getByText(/daily coding challenge not found\./i)
+    ).toBeVisible();
+  });
+
+  test('should show not found page for API error', async ({ page }) => {
+    await page.route(dateRouteRe, async route => {
+      await route.fulfill({
+        status: 500,
+        headers: { 'Content-Type': 'application/json' },
+        json: { type: 'error', message: 'Internal server error.' }
+      });
+    });
+
+    await page.goto('/learn/daily-coding-challenge?date=2025-01-01');
+    await expect(
+      page.getByText(/daily coding challenge not found\./i)
+    ).toBeVisible();
+  });
+
+  test('should show not found page for invalid challenge data', async ({
```

Comment:
It's not a deal-breaker, but I think it would be helpful if something other than the 404 page was shown in this case. It would make it easier to understand what's going wrong when users report issues.

Reply:
Thanks. Added to the list.